pulser-core/pulser/math/abstract_array.py

```diff
@@ -0,0 +1,301 @@
+# Copyright 2024 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the AbstractArray class."""
+from __future__ import annotations
+
+import functools
+import importlib.util
+import operator
+from typing import Any, Generator, Union, cast
+
+import numpy as np
+from numpy.typing import ArrayLike, DTypeLike
+
+from pulser.json.utils import obj_to_dict
+
+try:
+    import torch
+except ImportError:  # pragma: no cover
+    pass
+
+
+class AbstractArray:
+    """An abstract array containing an array or tensor.
+
+    Args:
+        array: The array to store.
+    """
+
+    def __init__(
+        self,
+        array: AbstractArrayLike,
+        dtype: DTypeLike = None,
```

Comment:
- `torch.Tensor` is actually `ArrayLike` so it is also `AbstractArrayLike`. 
- `DTypeLike` is a bit abusive but it works for the usual built-in types (eg `int`, `float`, `complex`, `bool`, ...). You'll see there is a `# type: ignore[arg-type]` further down because it doesn't fully work.